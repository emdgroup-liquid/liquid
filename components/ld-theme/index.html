
<!doctype html>
<html lang="en">
<head>
  <base href="/liquid/">
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Liquid â€“ Theme</title>

  <link rel="apple-touch-icon" sizes="180x180" href="/liquid/assets/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/liquid/assets/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/liquid/assets/favicon-16x16.png">
  <link rel="manifest" href="/liquid/assets/site.webmanifest">
  <link rel="mask-icon" href="/liquid/assets/safari-pinned-tab.svg" color="#eb3c96">
  <meta name="msapplication-TileColor" content="#eb3c96">
  <meta name="theme-color" content="#ffffff">

  <script>
    __docsSearchIndex__ = [{
              title: 'CSS vs. Web Components',
              tags: '',
              url: '/liquid/liquid/components-assets/',
              headings: 'Tweaking the bundler and adjusting the assets path',
              breadcrumbs: [
                
                'Liquid','CSS vs. Web Components']
            },{
              title: 'Components',
              tags: '',
              url: '/liquid/components/',
              headings: 'Status',
              breadcrumbs: [
                
                ]
            },{
              title: 'Background Cells',
              tags: '',
              url: '/liquid/components/ld-bg-cells/',
              headings: 'DefaultCell typesCustom position, size and colorsCSS VariablesProperties',
              breadcrumbs: [
                
                'Components','Background Cells']
            },{
              title: 'Button',
              tags: '',
              url: '/liquid/components/ld-button/',
              headings: 'ExamplesPrimaryDisabledSecondaryOn brand colorSecondary on brand colorGhostHighlightDangerSizeWith iconWith custom widthJustify contentMulti-lineAnchor buttonProgress buttonProperties',
              breadcrumbs: [
                
                'Components','Button']
            },{
              title: 'Checkbox',
              tags: '',
              url: '/liquid/components/ld-checkbox/',
              headings: 'ExamplesPrimaryDisabledDarkHighlightInvalidDangerWith labelWith label and input messageInput validationPropertiesDependenciesUsed byGraph',
              breadcrumbs: [
                
                'Components','Checkbox']
            },{
              title: 'Heading',
              tags: '',
              url: '/liquid/components/ld-heading/',
              headings: 'ExamplesWith heading levelWith deviating visual level b*With deviating visual level xh*With deviating visual level xb*Heading colorsProperties',
              breadcrumbs: [
                
                'Components','Heading']
            },{
              title: 'Icon',
              tags: '',
              url: '/liquid/components/ld-icon/',
              headings: 'Web ComponentDifferent sizesColorWith custom SVG iconAvailable iconsPropertiesSlots',
              breadcrumbs: [
                
                'Components','Icon']
            },{
              title: 'Input Message',
              tags: '',
              url: '/liquid/components/ld-input-message/',
              headings: 'ExamplesAs error messageAs info messageAs success messageProperties',
              breadcrumbs: [
                
                'Components','Input Message']
            },{
              title: 'Input',
              tags: '',
              url: '/liquid/components/ld-input/',
              headings: 'ExamplesDefaultDisabledDarkInvalidType dateType emailType fileType numberType passwordType searchType telType timeType urlMultiline (as textarea)SizeWith labelWith label and input messageWith slotted elementsInput validationPropertiesSlots',
              breadcrumbs: [
                
                'Components','Input']
            },{
              title: 'Label',
              tags: '',
              url: '/liquid/components/ld-label/',
              headings: 'ExampleSizePositionProperties',
              breadcrumbs: [
                
                'Components','Label']
            },{
              title: 'Loading Indicator',
              tags: '',
              url: '/liquid/components/ld-loading/',
              headings: 'CSS VariablesProperties',
              breadcrumbs: [
                
                'Components','Loading Indicator']
            },{
              title: 'Notification',
              tags: '',
              url: '/liquid/components/ld-notification/',
              headings: 'How it worksAccessibilityNotifications with interactive contentNotification hierarchyNotification timeoutPreventing a timeoutTimeout handling for queued notificationsNotification contentRedundant notifications handlingDismissing notificaitonsDismissing current notificaitonClearing all notificationsExamplesProperties',
              breadcrumbs: [
                
                'Components','Notification']
            },{
              title: 'Paragraph',
              tags: '',
              url: '/liquid/components/ld-paragraph/',
              headings: 'ExamplesDefault paragraphSizesProperties',
              breadcrumbs: [
                
                'Components','Paragraph']
            },{
              title: 'Radio Button',
              tags: '',
              url: '/liquid/components/ld-radio/',
              headings: 'ExamplesPrimaryDisabledDarkHighlightInvalidDangerWith labelWith label and input messageInput validationProperties',
              breadcrumbs: [
                
                'Components','Radio Button']
            },{
              title: 'Select',
              tags: '',
              url: '/liquid/components/ld-select/',
              headings: 'ExamplesSingle select modeMultiple select modeDisabledInvalidDetachedInlineGhostSizeWith custom trigger button iconWith labelWith label and input messageInput validationPropertiesEventsSlots',
              breadcrumbs: [
                
                'Components','Select']
            },{
              title: 'Screen Reader Live',
              tags: '',
              url: '/liquid/components/ld-sr-live/',
              headings: 'How it worksExampleProperties',
              breadcrumbs: [
                
                'Components','Screen Reader Live']
            },{
              title: 'Screen Reader Only',
              tags: '',
              url: '/liquid/components/ld-sr-only/',
              headings: 'ExamplesProperties',
              breadcrumbs: [
                
                'Components','Screen Reader Only']
            },{
              title: 'Theme',
              tags: '',
              url: '/liquid/components/ld-theme/',
              headings: 'How to apply a themeTheme inceptionProperties',
              breadcrumbs: [
                
                'Components','Theme']
            },{
              title: 'Toggle',
              tags: '',
              url: '/liquid/components/ld-toggle/',
              headings: 'ExamplesPrimaryLargeInvalidDisabledWith iconsWith icons (disabled)With icons (invalid)With labelWith label and input messageProperties',
              breadcrumbs: [
                
                'Components','Toggle']
            },{
              title: 'Tooltip',
              tags: '',
              url: '/liquid/components/ld-tooltip/',
              headings: 'DefaultWith arrowWith custom triggerOpen on clickPositioningShow/hide delayCSS VariablesProperties',
              breadcrumbs: [
                
                'Components','Tooltip']
            },{
              title: 'CSS vs. Web Components',
              tags: '',
              url: '/liquid/liquid/css-vs-web-components/',
              headings: 'When to use what',
              breadcrumbs: [
                
                'Liquid','CSS vs. Web Components']
            },{
              title: 'Getting started',
              tags: '',
              url: '/liquid/liquid/getting-started/',
              headings: 'Import and register Web Componentsself-lazy-loading bundlecustom-elements bundleImport stylesheets',
              breadcrumbs: [
                
                'Liquid','Getting started']
            },{
              title: 'Globals',
              tags: '',
              url: '/liquid/global/',
              headings: '',
              breadcrumbs: [
                
                ]
            },{
              title: 'Border-radius',
              tags: '',
              url: '/liquid/global/border-radius/',
              headings: '',
              breadcrumbs: [
                
                'Globals','Border-radius']
            },{
              title: 'Colors',
              tags: '',
              url: '/liquid/global/colors/',
              headings: 'White and Liquid BlackGrey scale colorsBackground grey scale colorsForeground grey scale colorsRich colorsRich BlueRich GreenRich PurpleRich RedVibrant colorsVibrant CyanVibrant GreenVibrant MagentaVibrant YellowTransparent colorsTheme palettesBubblegumOceanShakeSolventTea',
              breadcrumbs: [
                
                'Globals','Colors']
            },{
              title: 'Focus',
              tags: '',
              url: '/liquid/global/focus/',
              headings: '',
              breadcrumbs: [
                
                'Globals','Focus']
            },{
              title: 'Fonts',
              tags: '',
              url: '/liquid/global/fonts/',
              headings: '',
              breadcrumbs: [
                
                'Globals','Fonts']
            },{
              title: 'Shadows',
              tags: '',
              url: '/liquid/global/shadows/',
              headings: 'Box shadowsActiveHoverStackedStickyDrop-shadows (CSS filter)ActiveHoverStackedSticky',
              breadcrumbs: [
                
                'Globals','Shadows']
            },{
              title: 'Spacings',
              tags: '',
              url: '/liquid/global/spacings/',
              headings: '',
              breadcrumbs: [
                
                'Globals','Spacings']
            },{
              title: 'Typography',
              tags: '',
              url: '/liquid/global/typography/',
              headings: 'Available CSS custom properties',
              breadcrumbs: [
                
                'Globals','Typography']
            },{
              title: 'Server-side rendering',
              tags: '',
              url: '/liquid/liquid/server-side-rendering/',
              headings: 'Known issues and workarounds',
              breadcrumbs: [
                
                'Liquid','Server-side rendering']
            },{
              title: 'Type checking and intellisense',
              tags: '',
              url: '/liquid/liquid/type-checking-and-intellisense/',
              headings: 'With JSXWith vanilla HTML',
              breadcrumbs: [
                
                'Liquid','Type checking and intellisense']
            }
    ]
  </script>

  <link rel="stylesheet" href="/liquid/docs.css">
  <link rel="stylesheet" href="/liquid/dist/build/liquid.css">
  <script type="module" src="/liquid/dist/build/liquid.esm.js"></script>
  <script nomodule src="/liquid/dist/build/liquid.js"></script>
</head>
<body>
  <script>
    if (/Safari/.test(navigator.userAgent) && /Apple Computer/.test(navigator.vendor)) {
      document.body.classList.add('sarari')
    }

    var storedThemePref = window.localStorage.getItem('docs-theme')
    var isDark = storedThemePref === 'dark' ? true : false
    document.body.classList.add(`docs-theme-${this.isDark ? 'dark' : 'light'}`)
    document.body.classList.remove(
      `docs-theme-${!this.isDark ? 'dark' : 'light'}`
    )
  </script>
  <docs-layout id="docs-layout" style="visibility: hidden">
    <docs-topbar></docs-topbar>

    <docs-nav>
      
      

      <ul>
          <li class="docs-nav__li docs-nav__li--active"><details>
                <summary class="docs-nav__summary docs-nav__summary--liquid">
                  <a href="/liquid/"><span aria-label="Liquid">Liquid</span></a>
                </summary>
                <ul>
          <li class="docs-nav__li"><a href="/liquid/liquid/getting-started/">Getting started</a></li>
          <li class="docs-nav__li"><a href="/liquid/liquid/css-vs-web-components/">CSS vs. Web Components</a></li>
          <li class="docs-nav__li"><a href="/liquid/liquid/components-assets/">Component assets</a></li>
          <li class="docs-nav__li"><a href="/liquid/liquid/type-checking-and-intellisense/">Type checking and intellisense</a></li>
          <li class="docs-nav__li"><a href="/liquid/liquid/server-side-rendering/">Server-side rendering</a></li></ul>
              </details></li>
          <li class="docs-nav__li"><details>
                <summary class="docs-nav__summary docs-nav__summary--globals">
                  <a href="/liquid/global/"><span aria-label="Globals">Globals</span></a>
                </summary>
                <ul>
          <li class="docs-nav__li"><a href="/liquid/global/border-radius/">Border-radius</a></li>
          <li class="docs-nav__li"><a href="/liquid/global/colors/">Colors</a></li>
          <li class="docs-nav__li"><a href="/liquid/global/focus/">Focus</a></li>
          <li class="docs-nav__li"><a href="/liquid/global/fonts/">Fonts</a></li>
          <li class="docs-nav__li"><a href="/liquid/global/shadows/">Shadows</a></li>
          <li class="docs-nav__li"><a href="/liquid/global/spacings/">Spacings</a></li>
          <li class="docs-nav__li"><a href="/liquid/global/typography/">Typography</a></li></ul>
              </details></li>
          <li class="docs-nav__li docs-nav__li--active"><details open>
                <summary class="docs-nav__summary docs-nav__summary--components">
                  <a href="/liquid/components/"><span aria-label="Components">Components</span></a>
                </summary>
                <ul>
          <li class="docs-nav__li"><a href="/liquid/components/ld-bg-cells/">Background Cells</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-button/">Button</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-checkbox/">Checkbox</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-heading/">Heading</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-icon/">Icon</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-input/">Input</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-input-message/">Input Message</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-label/">Label</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-loading/">Loading Indicator</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-notification/">Notification</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-paragraph/">Paragraph</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-radio/">Radio Button</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-sr-live/">Screen Reader Live</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-sr-only/">Screen Reader Only</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-select/">Select</a></li>
          <li class="docs-nav__li docs-nav__li--active"><a href="/liquid/components/ld-theme/">Theme</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-toggle/">Toggle</a></li>
          <li class="docs-nav__li"><a href="/liquid/components/ld-tooltip/">Tooltip</a></li></ul>
              </details></li></ul>
    </docs-nav>

    <docs-nav-backdrop></docs-nav-backdrop>

    <docs-btn-toggle-nav></docs-btn-toggle-nav>

    
      <docs-toc style="order: 2;">
        <nav class="docs-toc__nav">
        <ol><li><a href="#how-to-apply-a-theme">How to apply a theme</a></li><li><a href="#theme-inception">Theme inception</a></li><li><a href="#properties">Properties</a></li></ol></nav>
      </docs-toc>
    

    <docs-main>
      <header><docs-breadcrumbs>
            
            <li>
                <a href="/liquid/components/">Components</a>
              </li><li>
              <a href="/liquid/components/ld-theme/" aria-current="page">Theme</a>
            </li>
          </docs-breadcrumbs><docs-edit-on-github path="./src/liquid/components/ld-theme/readme.md"></docs-edit-on-github>
      </header>
      <h1 id="ld-theme">ld-theme <a class="header-anchor" href="components/ld-theme/#ld-theme"><span aria-label="" class="header-anchor__symbol">#</span></a></h1>
<p>Liquid comes with multiple themes for theming its UI components. Most of the examples in the documentation have a theme switch built in, which you can use to switch between the available themes. In the following examples this switch is turned off. Instead, the examples show how you can apply a theme yourself.</p>
<hr>
<h2 id="how-to-apply-a-theme">How to apply a theme <a class="header-anchor" href="components/ld-theme/#how-to-apply-a-theme"><span aria-label="" class="header-anchor__symbol">#</span></a></h2>
<p>You apply a theme by wrapping whatever needs to be themed in an <code>ld-theme</code> component, or an element with a CSS theme class.</p>
<docs-example code="%3Cld-theme%20name%3D%22bubblegum%22%3E%0A%20%20%3Cld-button%3EText%3C%2Fld-button%3E%0A%3C%2Fld-theme%3E" code-css-component="%3Cdiv%20class%3D%22ld-theme-bubblegum%22%3E%0A%20%20%3Cbutton%20class%3D%22ld-button%22%3EText%3C%2Fbutton%3E%0A%3C%2Fdiv%3E"  opened>
<div slot="code">
<pre class="language-html"><code class="language-html"><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>ld-theme</span> <span class="token attr-name">name</span><span class="token attr-value"><span class="token punctuation attr-equals">=</span><span class="token punctuation">"</span>bubblegum<span class="token punctuation">"</span></span><span class="token punctuation">></span></span><br>  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>ld-button</span><span class="token punctuation">></span></span>Text<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>ld-button</span><span class="token punctuation">></span></span><br><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>ld-theme</span><span class="token punctuation">></span></span></code></pre>
</div><div slot="codeCssComponent">
<pre class="language-html"><code class="language-html"><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>div</span> <span class="token attr-name">class</span><span class="token attr-value"><span class="token punctuation attr-equals">=</span><span class="token punctuation">"</span>ld-theme-bubblegum<span class="token punctuation">"</span></span><span class="token punctuation">></span></span><br>  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>button</span> <span class="token attr-name">class</span><span class="token attr-value"><span class="token punctuation attr-equals">=</span><span class="token punctuation">"</span>ld-button<span class="token punctuation">"</span></span><span class="token punctuation">></span></span>Text<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>button</span><span class="token punctuation">></span></span><br><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>div</span><span class="token punctuation">></span></span></code></pre>
</div><div slot="show"><ld-theme name="bubblegum">
  <ld-button>Text</ld-button>
</ld-theme></div><div slot="showCssComponent"><div class="ld-theme-bubblegum">
  <button class="ld-button">Text</button>
</div></div></docs-example>
<h2 id="theme-inception">Theme inception <a class="header-anchor" href="components/ld-theme/#theme-inception"><span aria-label="" class="header-anchor__symbol">#</span></a></h2>
<p>In rare cases you will want to have a theming element wrapped by another theming element. Liquid supports a one level theme inception, which should be sufficiant for most edge cases. You can <strong>not</strong> wrap a theme in a theme in a theme... So, here is an example of a one level theme inception:</p>
<docs-example code="%3Cld-theme%20name%3D%22bubblegum%22%3E%0A%20%20%3Cld-theme%20name%3D%22tea%22%3E%0A%20%20%20%20%3Cld-button%3EText%3C%2Fld-button%3E%0A%20%20%3C%2Fld-theme%3E%0A%3C%2Fld-theme%3E" code-css-component="%3Cdiv%20class%3D%22ld-theme-bubblegum%22%3E%0A%20%20%3Cdiv%20class%3D%22ld-theme-tea%22%3E%0A%20%20%20%20%3Cbutton%20class%3D%22ld-button%22%3EText%3C%2Fbutton%3E%0A%20%20%3C%2Fdiv%3E%0A%3C%2Fdiv%3E"  opened>
<div slot="code">
<pre class="language-html"><code class="language-html"><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>ld-theme</span> <span class="token attr-name">name</span><span class="token attr-value"><span class="token punctuation attr-equals">=</span><span class="token punctuation">"</span>bubblegum<span class="token punctuation">"</span></span><span class="token punctuation">></span></span><br>  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>ld-theme</span> <span class="token attr-name">name</span><span class="token attr-value"><span class="token punctuation attr-equals">=</span><span class="token punctuation">"</span>tea<span class="token punctuation">"</span></span><span class="token punctuation">></span></span><br>    <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>ld-button</span><span class="token punctuation">></span></span>Text<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>ld-button</span><span class="token punctuation">></span></span><br>  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>ld-theme</span><span class="token punctuation">></span></span><br><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>ld-theme</span><span class="token punctuation">></span></span></code></pre>
</div><div slot="codeCssComponent">
<pre class="language-html"><code class="language-html"><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>div</span> <span class="token attr-name">class</span><span class="token attr-value"><span class="token punctuation attr-equals">=</span><span class="token punctuation">"</span>ld-theme-bubblegum<span class="token punctuation">"</span></span><span class="token punctuation">></span></span><br>  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>div</span> <span class="token attr-name">class</span><span class="token attr-value"><span class="token punctuation attr-equals">=</span><span class="token punctuation">"</span>ld-theme-tea<span class="token punctuation">"</span></span><span class="token punctuation">></span></span><br>    <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>button</span> <span class="token attr-name">class</span><span class="token attr-value"><span class="token punctuation attr-equals">=</span><span class="token punctuation">"</span>ld-button<span class="token punctuation">"</span></span><span class="token punctuation">></span></span>Text<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>button</span><span class="token punctuation">></span></span><br>  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>div</span><span class="token punctuation">></span></span><br><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>div</span><span class="token punctuation">></span></span></code></pre>
</div><div slot="show"><ld-theme name="bubblegum">
  <ld-theme name="tea">
    <ld-button>Text</ld-button>
  </ld-theme>
</ld-theme></div><div slot="showCssComponent"><div class="ld-theme-bubblegum">
  <div class="ld-theme-tea">
    <button class="ld-button">Text</button>
  </div>
</div></div></docs-example>
<!-- Auto Generated Below -->
<h2 id="properties">Properties <a class="header-anchor" href="components/ld-theme/#properties"><span aria-label="" class="header-anchor__symbol">#</span></a></h2>
<table>
<thead>
<tr>
<th>Property</th>
<th>Attribute</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>key</code></td>
<td><code>key</code></td>
<td>for tracking the node's identity when working with lists</td>
<td><code>string \| number</code></td>
<td><code>undefined</code></td>
</tr>
<tr>
<td><code>name</code></td>
<td><code>name</code></td>
<td>The theme name.</td>
<td><code>&quot;bubblegum&quot; \| &quot;ocean&quot; \| &quot;shake&quot; \| &quot;solvent&quot; \| &quot;tea&quot;</code></td>
<td><code>'ocean'</code></td>
</tr>
<tr>
<td><code>ref</code></td>
<td><code>ref</code></td>
<td>reference to component</td>
<td><code>any</code></td>
<td><code>undefined</code></td>
</tr>
</tbody>
</table>
<hr>
<p><em>Built with <a href="https://stenciljs.com/">StencilJS</a></em></p>

    </docs-main>

  </docs-layout>
  <docs-search></docs-search>
</body>
</html>
