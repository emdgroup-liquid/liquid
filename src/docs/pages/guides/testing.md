---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

