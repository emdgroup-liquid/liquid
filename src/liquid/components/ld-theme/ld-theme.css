/* .ld-theme-ocean -> default */
:root,
:host(.ld-theme-ocean),
.ld-theme-ocean {
  --ld-thm-accent: var(--ld-thm-ocean-accent);
  --ld-thm-bg-primary: var(--ld-thm-ocean-bg-primary);
  --ld-thm-bg-secondary: var(--ld-thm-ocean-bg-secondary);
  --ld-thm-highlight: var(--ld-thm-ocean-highlight);
}

:host(.ld-theme-bubblegum),
.ld-theme-bubblegum {
  --ld-thm-accent: var(--ld-thm-bubblegum-accent);
  --ld-thm-bg-primary: var(--ld-thm-bubblegum-bg-primary);
  --ld-thm-bg-secondary: var(--ld-thm-bubblegum-bg-secondary);
  --ld-thm-highlight: var(--ld-thm-bubblegum-highlight);
}

:host(.ld-theme-shake),
.ld-theme-shake {
  --ld-thm-accent: var(--ld-thm-shake-accent);
  --ld-thm-bg-primary: var(--ld-thm-shake-bg-primary);
  --ld-thm-bg-secondary: var(--ld-thm-shake-bg-secondary);
  --ld-thm-highlight: var(--ld-thm-shake-highlight);
}

:host(.ld-theme-solvent),
.ld-theme-solvent {
  --ld-thm-accent: var(--ld-thm-solvent-accent);
  --ld-thm-bg-primary: var(--ld-thm-solvent-bg-primary);
  --ld-thm-bg-secondary: var(--ld-thm-solvent-bg-secondary);
  --ld-thm-highlight: var(--ld-thm-solvent-highlight);
}

:host(.ld-theme-tea),
.ld-theme-tea {
  --ld-thm-accent: var(--ld-thm-tea-accent);
  --ld-thm-bg-primary: var(--ld-thm-tea-bg-primary);
  --ld-thm-bg-secondary: var(--ld-thm-tea-bg-secondary);
  --ld-thm-highlight: var(--ld-thm-tea-highlight);
}
