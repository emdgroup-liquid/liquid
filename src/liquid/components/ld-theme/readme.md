---
eleventyNavigation:
  key: Theme
  parent: Components
layout: layout.njk
title: Theme
permalink: liquid/components/ld-theme/
---

# Theme

Liquid comes with multiple themes for theming its UI components. Most of the examples in the documentation have a theme switch built in, which you can use to switch between the available themes. In the following examples this switch is turned off. Instead, the examples show how you can apply a theme yourself.

## How to apply a theme

You apply a theme by wrapping whatever needs to be themed in an `ld-theme` component, or an element with a CSS theme class.

{% example 'html', false, false, false, false %}
<ld-theme name="bubblegum">
  <ld-button>Text</ld-button>
</ld-theme>

<!-- CSS component -->

<div class="ld-theme-bubblegum">
  <button class="ld-button">Text</button>
</div>
{% endexample %}

## Theme inception

In rare cases you will want to have a theming element wrapped by another theming element. Liquid supports a one level theme inception, which should be sufficiant for most edge cases. You can **not** wrap a theme in a theme in a theme... So, here is an example of a one level theme inception:

{% example 'html', false, false, false, false %}
<ld-theme name="bubblegum">
  <ld-theme name="tea">
    <ld-button>Text</ld-button>
  </ld-theme>
</ld-theme>

<!-- CSS component -->

<div class="ld-theme-bubblegum">
  <div class="ld-theme-tea">
    <button class="ld-button">Text</button>
  </div>
</div>
{% endexample %}


<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description     | Type                                                      | Default   |
| -------- | --------- | --------------- | --------------------------------------------------------- | --------- |
| `name`   | `name`    | The theme name. | `"bubblegum" \| "ocean" \| "shake" \| "solvent" \| "tea"` | `'ocean'` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
