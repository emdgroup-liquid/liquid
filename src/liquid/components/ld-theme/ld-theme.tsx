import '../../components' // type definitions for type checks and intelliSense
import { Component, h, Host, Prop } from '@stencil/core'

@Component({
  tag: 'ld-theme',
  styleUrl: 'ld-theme.css',
  shadow: false,
})
export class LdTheme {
  /** The theme name. */
  @Prop() name: 'ocean' | 'solvent' | 'bubblegum' | 'shake' | 'tea' = 'ocean'

  render() {
    return (
      <Host class={`ld-theme-${this.name}`}>
        <slot></slot>
      </Host>
    )
  }
}
