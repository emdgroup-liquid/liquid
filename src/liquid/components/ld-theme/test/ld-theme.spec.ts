import { newSpecPage } from '@stencil/core/testing'
import { LdTheme } from '../ld-theme'

describe('ld-theme', () => {
  it('renders default theme', async () => {
    const page = await newSpecPage({
      components: [LdTheme],
      html: `<ld-theme>yolo</ld-theme>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders explicit theme', async () => {
    const page = await newSpecPage({
      components: [LdTheme],
      html: `<ld-theme name="tea">yolo</ld-theme>`,
    })
    expect(page.root).toMatchSnapshot()
  })
})
