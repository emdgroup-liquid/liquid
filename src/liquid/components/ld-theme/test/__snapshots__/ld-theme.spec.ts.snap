// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-theme renders default theme 1`] = `
<ld-theme class="ld-theme-ocean">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  yolo
</ld-theme>
`;

exports[`ld-theme renders explicit theme 1`] = `
<ld-theme class="ld-theme-tea" name="tea">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  yolo
</ld-theme>
`;
